src/ParticleDistributions/ParticleDistributions.jl

```diff
@@ -633,49 +711,86 @@ parameters (x) of the given distribution
-  F = []
-  # M0 (==M[1]) is the sum of all n_j: M[1] = ∑n_j
-  F_0 = M[1] - sum(x[1:m])
-  push!(F, F_0)
-  for i in 1:(n_equations-1)
-    F_i = M[i+1]
-    for j in 1:m
-      # x[j+2*m] is k_j
-      factor = prod([(x[j+2*m]+counter) for counter in 0:(i-1)])
-      # x[j] is n_j,  x[j+m] is θ_j
-      F_i -= x[j] * x[j+m]^i * factor
-    end
-    push!(F, F_i)
-  end
+  p1 = ones(FT, 1, n_equations)
+  p2 = collect(reshape(0:n_equations-1, 1, n_equations))
+  k_prod(k, n) = (n == 0) ? 1 : k_prod(k, n-1) * (k + n-1)
+  g(n, θ, k) = sum(n.^p1 .* θ.^p2 .* k_prod.(k, p2), dims = 1)[:]
 
-  return F
+  return log.(M) - log.(g(x[1:3:end], x[2:3:end], x[3:3:end]))
 end
 
 function construct_system(x, dist::ExponentialAdditiveParticleDistribution{FT}, M::Array{FT}) where {FT<:Real}
   n_equations = length(M)
-  m = length(dist.subdists)
 
-  # x: [n_1, ..., n_m, θ_1, ..., θ_m]
+  # x: [n_1, θ_1, ..., n_m, θ_m]
   # Construct system of n_equations equations relating moments to parameters
-  # M[i] = i! * ∑(n[j] * θ[j]^i) 
-  F = []
-  # M0 (==M[1]) is the sum of all n_j: M[1] = ∑n_j
-  F_0 = M[1] - sum(x[1:m])
-  push!(F, F_0)
-  for i in 1:(n_equations-1)
-    # x[j] is n_j, x[j+m] is θ_j
-    F_i = M[i+1] - factorial(i) * sum([x[j] * x[j+m]^i for j in 1:m])
-    push!(F, F_i)
-  end
-
-  return F
+  # M_i = i! * ∑(n[j] * θ[j]^i)
+  c1 = factorial.(0:n_equations-1)
+  p1 = ones(FT, 1, n_equations)
+  p2 = collect(reshape(0:n_equations-1, 1, n_equations))
+  g(n, θ) = c1 .* sum(n.^p1 .* θ.^p2, dims = 1)[:]
+
+  return M - g(x[1:2:end], x[2:2:end])
+end
+
+function construct_system(x, dist::MonodisperseAdditiveParticleDistribution{FT}, M::Array{FT}) where {FT<:Real}
+  n_equations = length(M)
+
+  # x: [n_1, θ_1, ..., n_m, θ_m]
+  # Construct system of n_equations equations relating moments to parameters
+  # M_i = ∑(n[j] * θ[j]^i) 
+  p1 = ones(FT, 1, n_equations)
+  p2 = collect(reshape(0:n_equations-1, 1, n_equations))
+  g(n, θ) = sum(n.^p1 .* θ.^p2, dims = 1)[:]
+    
+  return M - g(x[1:2:end], x[2:2:end])
+end
+
+"""
+  moment_source_helper(dist, a, b, x_star)
+
+  - `dist` - AbstractParticleDistribution
+  - `p1` - power of particle mass
+  - `p2` - power of particle mass
+  - `x_star`- particle mass threshold
+
+Returns ∫_0^x_star ∫_0^(x_star-x') x^p1 x'^p2 f(x) f(x') dx dx' for computations of the source of moments of the distribution below
+the given threshold x_star.
+"""
+function moment_source_helper(dist::MonodispersePrimitiveParticleDistribution{FT}, p1::FT, p2::FT, x_star::FT) where {FT<:Real}
+  n, θ = get_params(dist)[2]
+  source = (θ < x_star/2) ? n^2 * θ^(p1+p2) : 0
+  return source
+end
+
+function moment_source_helper(dist::ExponentialPrimitiveParticleDistribution{FT}, p1::FT, p2::FT, x_star::FT) where {FT<:Real}
+  n, θ = get_params(dist)[2]
+
+  f(x) = x^p1 * exp(-x/θ) * gamma_inc(p2 + 1, (x_star - x)/θ)[1] * gamma(p2 + 1)
+  
+  logx = range(log(1e-5), log(x_star), 51)
```

Comment:
Let's specify this lower bound as an optional argument; maybe also the number of points sampled?